packages/starksheet-webapp/src/provider/EVMProvider.ts

```diff
@@ -190,31 +192,75 @@ export class EVMProvider implements ChainProvider {
   execute = async (calls: Call[], options: { value: number | string })  => {
     const provider = new ethers.providers.Web3Provider(window.ethereum);
     const signer = provider.getSigner();
+    const MULTISEND_CONTRACT_ADDRESS = process.env.REACT_APP_MULTISEND_ADDRESS || "";
 
+    const encodeTransactions = await Promise.all(calls.map(async (call, index) => {
+      let abi = await this.getAbi(call.contractAddress);
+      if(abi.length === 0) {
+        abi = evmWorksheetAbi;
+      }
+      // todo: implement cleanly
 
-    // Callers of the execute function assume the chain supports multi-call, which EVM does not natively.
-    // The following is a hack: we wait for every transaction before resolving, then returning one of the transactions.
-    // This works because callers do not really care about the transaction response, only its status and completion
-    const receipts = await Promise.all(calls.map(async (call) => {
-      const abi = await this.getAbi(call.contractAddress);
       const contract = new Contract(call.contractAddress, abi, signer);
 
-      const response: ethers.providers.TransactionResponse = await contract[call.entrypoint](...(call.calldata || []), {
-        value: options.value,
-      });
-      return await response.wait();
-      }));
+      const fragment = contract.interface.fragments.find((fragment) => fragment.name === call.entrypoint);
 
-    const firstToFail = receipts.find((receipt) => receipt.status !== 1)
-    if (firstToFail) {
-      return {
-        transaction_hash: firstToFail.transactionHash,
+      if (!fragment) {
+        throw new Error(`Could not find fragment for ${call.entrypoint} entrypoint in contract ABI`);
+      }
+      const signature = fragment.format();
+  
+      const inputValues= {
+        0: ""
       };
+      (call.calldata as RawCalldata).forEach((value, index) => {
```

Comment:
this can be rewritten more cleanly with a reduce. Generally speaking, here is an example:
```
Object.keys(call.calldata).reduce((acc, key, index) => ({ ...acc, [index]: call.calldata[key] }), {})
```

you shouldn't use foreach for side-effects (e.g. mutating/building an outer scope variable).

That being said, looking at what the code achieves, if you just need to turn an array into an object with keys being indexes (0,1,2à, you can simply do a `const inputValues = Object.assign({}, call.calldata)`